File: cpp-alg/leecode/Q206_reverse_linked_list.cpp
#include <iostream>

struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

ListNode *reverseLinkedList(ListNode *head) {
  ListNode *prev = nullptr;
  ListNode *current = head;
  while (current) {
    ListNode *next = current->next;
    current->next = prev;
    prev = current;
    current = next;
  }
  return prev;
}

int main() {
  ListNode *nodeE = new ListNode(5);
  ListNode *nodeD = new ListNode(4, nodeE);
  ListNode *nodeC = new ListNode(3, nodeD);
  ListNode *nodeB = new ListNode(2, nodeC);
  ListNode *nodeA = new ListNode(1, nodeB);

  ListNode *result = reverseLinkedList(nodeA);
  while (result) {
    std::cout << result->val << " ";
    result = result->next;
  }
  std::cout << std::endl;
}

File: python-alg/leecode/q206_reverse_linked_list.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


def reverse_linked_list(node):
    if node is None:
        return
    prev = None
    curr = node
    while curr:
        temp = curr.next
        curr.next = prev
        prev = curr
        curr = temp
    return prev


nodeE = ListNode(5)
nodeD = ListNode(4)
nodeD.next = nodeE
nodeC = ListNode(3)
nodeC.next = nodeD
nodeB = ListNode(2)
nodeB.next = nodeC
nodeA = ListNode(1)
nodeA.next = nodeB

root = reverse_linked_list(nodeA)
while root:
    print(root.val)
    root = root.next
